addons/mail/tests/test_mail_tools.py

```diff
@@ -176,6 +176,36 @@ def test_mail_find_partner_from_emails_followers(self):
         found = self.env['res.partner']._mail_find_partner_from_emails([self.env.user.partner_id.email_formatted])
         self.assertEqual(found, [self.env.user.partner_id])
 
+    def test_mail_find_partner_from_emails_multicompany(self):
+        """ Test _mail_find_partner_from_emails when dealing with records in
+        a multicompany environment, returning a partner record with matching
+        company_id. """
+        self._activate_multi_company()
+        Partner = self.env['res.partner']
+        self.test_partner.company_id = self.company_2
+
+        # mixing up id ordering (p2.id < p1.id < p3.id)
+        test_partner_1 = self.test_partner.copy({'company_id': False})
+        test_partner_2 = self.test_partner
+        test_partner_3 = test_partner_1.copy({'company_id': self.company_3.id})
+        records = [
+            None,
+            *Partner.create([
+                {'name': 'Company 2 contact', 'company_id': self.company_2.id},
+                {'name': 'Company 3 contact', 'company_id': self.company_3.id},
+                {'name': 'No restrictions', 'company_id': False},
+            ])
+        ]
+        expected_partners = [
+            test_partner_1,
+            test_partner_2,
+            test_partner_3,
+            test_partner_1,
+        ]
+        for record, expected_partner in zip(records, expected_partners):
+            found = Partner._mail_find_partner_from_emails([self._test_email], record)
```

Comment:
As records is a named parameter, I prefer to use it like it
```suggestion
            found = Partner._mail_find_partner_from_emails([self._test_email], records=record)
```